### ISSUE TITLE
grunt fails to complete build

### ISSUE THREAD

**Issue by afshin2003** [noflo/noflo-ui#510]
Hi,
I do not know if something is wrong with my network or not!

BEST,
.
.
.
Running "noflo_browser" task
Running "noflo_browser:main" (noflo_browser) task
Verifying property noflo_browser.main exists in config...OK
Files: component.json -> browser/noflo-ui.js
Options: require, development=false, debug=false, ide="https://app.flowhub.io", signalserver="https://api.flowhub.io", concurrency=10, plugins=["component-json","component-coffee","component-fbp"], remotes=[], graph_scripts=[], heads=[]
Reading /opt/test/noflo-ui/node_modules/grunt-noflo-browser/templates/graph.html...OK
install noflo/noflo@master
install noflo/noflo-strings@master
install noflo/noflo-ajax@master
install noflo/noflo-localstorage@master
install noflo/noflo-interaction@master
install noflo/noflo-objects@master
install noflo/noflo-groups@master
install noflo/noflo-routers@master
install noflo/noflo-dom@master
install noflo/noflo-core@master
install noflo/noflo-polymer@master
install noflo/noflo-indexeddb@master
install noflo/noflo-github@master
install noflo/noflo-graph@master
install noflo/noflo-runtime@master
install noflo/noflo-runtime-webrtc@master
install the-grid/flowhub-registry@master
install jonnor/microflo@master
install gjohnson/uuid@master
install bergie/octo@master
install visionmedia/superagent@0.21.0
install jashkenas/underscore@master
install flowbased/fbp-spec@master
exists noflo/noflo-graph@master
Warning: connect ETIMEDOUT Use --force to continue.

Aborted due to warnings.

**Comment by rogsmith** [comment]
Did you ever get the build do work? When I run grunt build it fails with: TypeError: key must be a string or number. undefined in lru-cache.js:18:11

**Comment by jonnor** [comment]
Which versions of nodejs and NPM are you using? The build works on Travis (Saucelab tests sometimes fails though). https://travis-ci.org/noflo/noflo-ui/builds/94453497

**Comment by afshin2003** [comment]
...
install flowbased/fbp-spec@master
Warning: connect ECONNREFUSED 10.10.34.36:443 Use --force to continue.

**Comment by jonnor** [comment]
Please attach full log from clean rebuild (after say `git clean -dfx`).

Do you use a proxy?

**Comment by afshin2003** [comment]
Hi,
No I don't use proxy!
[grunt-build.txt](https://github.com/noflo/noflo-ui/files/69766/grunt-build.txt)

**Comment by afshin2003** [comment]
root@biz1:/# node -v
v4.2.3
root@biz1:/# npm -v
2.14.7

**Comment by afshin2003** [comment]
Hi,
Something should be wrong in the site that I am trying to install. That IP address is a private one. Probably a router / firewall. The thing that I can't understand is how the build process has acquired it and tries to connect!

**Comment by AxelVoitier** [comment]
Hi,

I also have various failed. Some random due to what seemed a temp issue with github CDN or maybe grunt failing randomly its network connections? After retrying couple of times it ends up getting the files it was missing before.

However I have this error that is related to [this](https://github.com/jashkenas/underscore/commit/8f3d0507e885f64a05424fdc09a07cf77e2f676b) commit in underscore repo:
```
Step 13 : RUN grunt build
 ---> Running in 76dc0957efa5
Running "inlinelint:all" (inlinelint) task
>> 44 files lint free.

Running "noflo_manifest:update" (noflo_manifest) task
File "component.json" updated.

Running "bower-install-simple:deps" (bower-install-simple) task
Executing Bower (Command: install)

Running "noflo_browser:main" (noflo_browser) task
install noflo/noflo@master
install noflo/noflo-strings@master
install noflo/noflo-ajax@master
install noflo/noflo-localstorage@master
install noflo/noflo-interaction@master
install noflo/noflo-objects@master
install noflo/noflo-groups@master
install noflo/noflo-routers@master
install noflo/noflo-dom@master
install noflo/noflo-core@master
install noflo/noflo-polymer@master
install noflo/noflo-indexeddb@master
install noflo/noflo-github@master
install noflo/noflo-graph@master
install noflo/noflo-runtime@master
install noflo/noflo-runtime-webrtc@master
install the-grid/flowhub-registry@master
install jonnor/microflo@master
install gjohnson/uuid@master
install bergie/octo@master
install visionmedia/superagent@0.21.0
install jashkenas/underscore@master
install flowbased/fbp-spec@master
Warning: failed to fetch https://raw.githubusercontent.com/jashkenas/underscore/master/component.json, got 404 "Not Found" Use --force to continue.

Aborted due to warnings.
Removing intermediate container 76dc0957efa5
The command '/bin/sh -c grunt build' returned a non-zero code: 6
```

If I install underscore beforehand with `npm install underscore` and then use `grunt build --force` I now have:
```
Step 13 : RUN grunt build --force
 ---> Running in 06e8e4d83eb8
Running "inlinelint:all" (inlinelint) task
>> 44 files lint free.

Running "noflo_manifest:update" (noflo_manifest) task
File "component.json" updated.

Running "bower-install-simple:deps" (bower-install-simple) task
Executing Bower (Command: install)

Running "noflo_browser:main" (noflo_browser) task
install noflo/noflo@master
install noflo/noflo-strings@master
install noflo/noflo-ajax@master
install noflo/noflo-localstorage@master
install noflo/noflo-interaction@master
install noflo/noflo-objects@master
install noflo/noflo-groups@master
install noflo/noflo-routers@master
install noflo/noflo-dom@master
install noflo/noflo-core@master
install noflo/noflo-polymer@master
install noflo/noflo-indexeddb@master
install noflo/noflo-github@master
install noflo/noflo-graph@master
install noflo/noflo-runtime@master
install noflo/noflo-runtime-webrtc@master
install the-grid/flowhub-registry@master
install jonnor/microflo@master
install gjohnson/uuid@master
install bergie/octo@master
install visionmedia/superagent@0.21.0
install jashkenas/underscore@master
install flowbased/fbp-spec@master
Warning: failed to fetch https://raw.githubusercontent.com/jashkenas/underscore/master/component.json, got 404 "Not Found" Used --force, continuing.
>> Error: can't find remote for "jashkenas/underscore"
the-grid/flowhub-registry complete
noflo/noflo-ajax complete
bergie/octo complete
gjohnson/uuid complete
noflo/noflo-core complete
visionmedia/superagent complete
noflo/noflo-github complete
noflo/noflo-groups complete
noflo/noflo-indexeddb complete
noflo/noflo-strings complete
noflo/noflo-routers complete
noflo/noflo-graph complete
noflo/noflo-localstorage complete
noflo/noflo complete
noflo/noflo-dom complete
jonnor/microflo complete
noflo/noflo-objects complete
noflo/noflo-interaction complete
noflo/noflo-polymer complete
noflo/noflo-runtime-webrtc complete
noflo/noflo-runtime complete
flowbased/fbp-spec complete
 ---> 5de429b31343
Removing intermediate container 06e8e4d83eb8
```

So, it seems it completed (but why does it keep trying to get underscore despite being already installed?!).
But then I don't have any index.html file (it is supposed to be generated, right?):
```
# ls -la
total 136
drwxr-xr-x 14 root root  4096 Jan  7 18:27 .
drwxr-xr-x 11 root root  4096 Jan  7 18:27 ..
drwxrwxr-x  8 root root  4096 Jan  7 17:31 .git
-rw-rw-r--  1 root root   297 Jan  7 17:31 .gitignore
-rw-rw-r--  1 root root   901 Jan  7 17:31 .travis.yml
-rw-rw-r--  1 root root  7270 Jan  7 17:31 CHANGES.md
-rw-rw-r--  1 root root   680 Jan  7 18:22 Dockerfile
-rw-rw-r--  1 root root 12092 Jan  7 17:31 Gruntfile.coffee
-rw-rw-r--  1 root root  1072 Aug 18  2014 LICENSE-MIT
-rw-rw-r--  1 root root  3494 Jan  7 17:31 README.md
drwxrwxr-x  2 root root  4096 Jan  7 17:31 app
-rw-rw-r--  1 root root   771 Jan  7 17:31 bower.json
drwxr-xr-x 20 root root  4096 Jan  7 18:26 bower_components
-rw-rw-r--  1 root root  7367 Jan  7 18:27 component.json
drwxrwxr-x 44 root root  4096 Jan  7 18:28 components
-rw-rw-r--  1 root root  1460 Aug 18  2014 config.dist.xml
drwxrwxr-x  2 root root  4096 Jan  7 17:31 css
drwxrwxr-x  2 root root  4096 Jan  7 17:31 elements
-rw-rw-r--  1 root root   147 Aug 18  2014 examples.json
drwxrwxr-x  2 root root  4096 Jan  7 17:31 graphs
-rw-rw-r--  1 root root  4173 Jan  7 17:31 index.dist.html
-rw-rw-r--  1 root root  1104 Jan  7 17:31 manifest.dist.json
-rw-rw-r--  1 root root  1125 Jan  7 17:31 manifest.dist.webapp.json
drwxr-xr-x 27 root root  4096 Jan  7 18:26 node_modules
-rw-rw-r--  1 root root  4286 Aug 18  2014 noflo.ico
-rw-rw-r--  1 root root  1225 Jan  7 17:31 package.json
drwxrwxr-x  2 root root  4096 Jan  7 17:31 spec
drwxrwxr-x  2 root root  4096 Aug 18  2014 src
```

**Comment by AxelVoitier** [comment]
Forgot the versions:
```
# node --version
v4.2.4
root@e60df4449e06:/# npm --version
2.14.12
```

**Comment by AxelVoitier** [comment]
Well, looking at the output of your CI build, it seems in my case `grunt build` does not execute the remaining tasks.

But if I execute them manually I get the following:
```
# grunt copy:themes
Running "copy:themes" (copy) task
Copied 2 files

Done, without errors.
root@f8d0c7a1a5ce:/usr/src/app# grunt vulcanize:app
Running "vulcanize:app" (vulcanize) task
Warning: ENOENT: no such file or directory, open '/usr/src/app/browser/noflo-noflo-polymer/noflo-polymer/noflo-polymer.html' Use --force to continue.

Aborted due to warnings.
```

Can't do much from here, the "browser" folder really do not exists...

**Comment by jonnor** [comment]
That `underscore` issue is different from that by original reporter here, it is caused by a recent upstream change. Propose we keep that in https://github.com/noflo/noflo/issues/355

**Comment by AxelVoitier** [comment]
Yes, I initially though it was linked but after further investigation I opened noflo/noflo#355, indeed ;).

**Comment by bergie** [comment]
We're now using webpack and npm for deps, so this should no longer be an issue

**Comment by Mrinaldo** [comment]
I am getting this error while doing 
`npm run build`

> Running "exec:vulcanize" (exec) task
>> 'node_modules' is not recognized as an internal or external command,
>> operable program or batch file.
>> Exited with code: 1.
>> Error executing child process: Error: Process exited with code 1.
Warning: Task "exec:vulcanize" failed. Use --force to continue.

**Comment by hpenav** [comment]
@Mrinaldo 
I got the same issue trying npm run build.... it seems is related on how windows OS handles the following command: 'node_modules/.bin/polymer-bundler index.dist.html > index.html' inside Gruntfile.coffe
Here is what I did in order to make it work

1. Open Gruntfile.coffe file and search for vulcanize
2. Replace command: 'node_modules/.bin/polymer-bundler index.dist.html > index.html' with
   command: 'vulcanize.bat'
3. Create vulcanize.bat file in the same folder where Gruntfile.coffe resides including the following line
  ./node_modules/.bin/polymer-bundler index.dist.html > index.html
  (it should include ./  prefix)
  save vulcanize.bat
4. Run again: npm run build

I did above and I got my noflo ui working
I  think the issue is related with windows OS having problems to execute commands the way Gruntfile.coffe implements them, I assume build process will work just fine on any linux OS... just guessing

Hope this helps you moving forward
